openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,563 @@
+                'data_account_type_prepayments',
+                'data_account_type_fixed_assets',
+                'data_account_type_current_liabilities',
+                'data_account_type_non_current_liabilities',
+                'data_account_type_equity',
+                'data_unaffected_earnings',
+                'data_account_type_revenue',
+                'data_account_type_other_income',
+                'data_account_type_expenses',
+                'data_account_type_depreciation',
+                'data_account_type_direct_costs',
+                'data_account_off_sheet'
+            )
+        )
+        UPDATE {table} aa
+            SET account_type = atm.account_type
+        FROM account_type_map atm
+        WHERE atm.user_type_id = aa.user_type_id
+        """,
+    )
+
+
+def _account_account_fast_fill_include_initial_balance(env):
+    if not openupgrade.column_exists(
+        env.cr, "account_account", "include_initial_balance"
+    ):
+        openupgrade.add_fields(
+            env,
+            [
+                (
+                    "include_initial_balance",
+                    "account.account",
+                    "account_account",
+                    "boolean",
+                    False,
+                    "account",
+                ),
+            ],
+        )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_account
+            SET include_initial_balance = true
+        WHERE account_type NOT IN
+        ('income', 'income_other', 'expense',
+        'expense_depreciation', 'expense_direct_cost', 'off_balance')
+        """,
+    )
+
+
+def _delete_sql_constraints(env):
+    # Delete constraints to recreate it
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_journal", "code_company_uniq"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_accountable_required_fields"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_amount_currency_balance_sign"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_credit_debit"
+    )
+
+
+def _account_analytic_line_fast_fill_journal_id(env):
+    if not openupgrade.column_exists(env.cr, "account_analytic_line", "journal_id"):
+        openupgrade.add_fields(
+            env,
+            [
+                (
+                    "journal_id",
+                    "account.analytic.line",
+                    "account_analytic_line",
+                    "many2one",
+                    False,
+                    "account",
+                )
+            ],
+        )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_analytic_line aal
+            SET journal_id = aml.journal_id
+        FROM account_move_line aml
+        WHERE aml.id = aal.move_line_id
+        """,
```

Comment:
I got `insert or update on table "account_analytic_line" violates foreign key constraint "account_analytic_line_journal_id_fkey"` on this update. 
We should probably drop the foreign key constraint before applying this.

Reply:
Another issue I've encountered around here is with analytic lines linked to the Timesheets Journal, which prevent the journal_id foreign key to apply. I've set journal_id to NULL on these records as a workaround. Not sure where the best place to do that is.